lms/services/roster.py

```diff
@@ -316,6 +326,127 @@ def fetch_canvas_group_roster(self, canvas_group: LMSSegment) -> None:
+             - We fetch them via the propietary Canvas API, not the LTI Names and Roles endpoint.
+
+             - Due to the return value of that API we don't fetch rosters for indivual sections,
+               but for all sections of one course at once
+
+             - The return value of the API doesn't include enought information to create unseen users
+               so we'll only match against users we have seen before in the course.
+        """
+        application_instance = self._get_application_instance(lms_course)
+
+        # Last instructor to launch the course, we'll use this user's API token to fetch the sections.
+        instructor = self._get_course_instructor(lms_course)
+        if not instructor:
+            LOG.info(
+                "Can't fetch roster for sections of course ID:%s. No instructor found.",
+                lms_course.id,
+            )
+            return
+
+        # Get all the sections for this course that we seen in the DB.
+        db_sections = self._db.scalars(
+            select(LMSSegment).where(
+                LMSSegment.lms_course_id == lms_course.id,
+                LMSSegment.type == "canvas_section",
+            )
+        ).all()
+
+        if not db_sections:
+            LOG.info(
+                "Can't fetch roster for sections of course ID:%s. No sections found in the DB.",
+                lms_course.id,
+            )
+            return
+
+        # We'll create a new Canvas API client for the relevant install and instructor to fetch the sections.
+        canvas_service = canvas_api_client_factory(
+            None,
+            self._request,
+            application_instance=application_instance,
+            user_id=instructor.lti_user_id,
+        )
+        # We'll take the token service from the client to refresh the token if needed.
+        # This is already scoped to the user and the install.
+        oauth2_token_service = canvas_service._client._oauth2_token_service  # noqa: SLF001
+
+        # Fetch the sections and their students from the Canvas API.
+        api_sections = self._get_canvas_sections(
+            canvas_service, oauth2_token_service, lms_course, with_refresh_token=True
+        )
+        if not api_sections:
+            LOG.info(
+                "Can't fetch roster for sections of course ID:%s. No sections found on the API.",
+                lms_course.id,
+            )
+            return
+        api_sections_by_id = {str(section["id"]): section for section in api_sections}
+
+        # The API doesn't send a LTI role, we'll pick a student one from the DB and use that
+        student_lti_role_id = self._db.scalar(
+            select(LTIRole.id)
+            .where(LTIRole.type == RoleType.LEARNER, LTIRole.scope == RoleScope.COURSE)
+            .order_by(LTIRole.id.asc())
+        )
+
+        roster_upsert_elements = []
+        db_course_users_by_lms_api_id = self._get_course_users(lms_course)
+        for db_section in db_sections:
```

Comment:
The overall process is:

- For each section we have in the DB
- Match it to the one of the sections from the API
- For each of the students from the API
- For the ones we match against one of the students in the course
- Include them in the roster.